Cython/Compiler/TypeStubGenerator.py

```diff
@@ -0,0 +1,433 @@
+    # the place of two of them I could see what Scoder meant when 
+    # said the original pull request needed to be cleaned up...
+
+    
+    def write_class(self, node, class_name):
+        self.endline()
+        self.put("class %s" % class_name)
+        if getattr(node,"bases",None) and isinstance(node.bases, TupleNode) and node.bases.args:
+            self.put("(")
+            self.put(",".join([name.name for name in node.bases.args]))
+            self.endline("):")
+        else:
+            self.endline(":")
+        self.class_func_count = 0
+        self._visitchildren_indented(node)
+        if self.class_func_count < 1:
+            self.indent()
+            self.putline("pass")
+            self.dedent()
+        self.class_func_count = 0
+        self.emptyline()
+        return node 
+    
+    # I have tried to merege these before via visit_ClassDefNode but it causes the system to break so this 
+    # was the best I could do to minigate the problem - Vizonex 
+    def visit_CClassDefNode(self, node):
+        return self.write_class(node, node.class_name)
+
+    def visit_PyClassDefNode(self, node):
+        return self.write_class(node, node.name)
+
+    def visit_CFuncDefNode(self, node):
+        # cdefs are for C only...
+        if not node.overridable:
+            return node 
+        
+        func_args = []
+        for arg in node.declarator.args:
+            value = ""
+            if not arg.declarator.name:
+                value = arg.base_type.name
+            elif hasattr(arg.base_type, "name"):
+                value = "%s : %s" % (arg.declarator.name, self.translate_base_type_to_py(arg.base_type))
+            if (arg.default is not None or
+                arg.default_value is not None):
+                value += " = ..."
+            func_args.append(value)
+        
+        self.class_func_count += 1
+        
+        func_name = node.declared_name()
+        self.startline()
+        self.put("def %s(" % func_name)
+    
+        self.put(", ".join(func_args))
+        
+        # TODO Maybe Try passing docstrings in the future for vscode users' sake
+        # or have it also be a compiler argument?...
+      
+        self.endline(") -> %s: ..." % self.translate_base_type_to_py(node.base_type))
+        
+        return node
+
+    
+    def write_decorator(self, decorator):
+        if isinstance(decorator, CallNode):
+            return
+        
+        self.startline("@")
+        if isinstance(decorator, NameNode):
+            self.endline("%s" % decorator.name)
+        else:
+            assert isinstance(decorator, AttributeNode) , "Decorator was not an attribute node..."
+            self.endline("%s.%s" % (decorator.obj.name,decorator.attribute))
+        
+
+    def annotation_Str(self, annotation):
+        return annotation.name if hasattr(annotation,"name") and annotation.is_name else  annotation.string.unicode_value 
```

Comment:
just `annotation.is_name` - I don't think you need to do `hasattr` and `is_name`

Reply:
Recursion again.